File: backend/app/services/issues/issue_links.py
"""The one builder of an issue's deep link (3a Task 3b).

The frontend route is ``/team/{team_id}/todolist/{issue_key}`` — keyed by the
issue KEY (``MH-n``) inside a team, never by the snowflake id. Two readers print
this URL today:

* the Generated inbox card's source line (``services/library/generated_source``)
* every version of a lineage response (``services/deliverables/lineage_view``)

They describe the same row from two directions, so they must print the same
string. A second builder would drift by a query string and nothing would fail —
the link would simply land at the top of the issue instead of on the step that
produced the thing you clicked from.

**It refuses more often than it builds.** No team or no key → ``None``, and the
caller renders a disabled control. That is the honest answer: a URL assembled
from ``issue_id`` alone resolves to nothing, and a dead link reads worse than a
button that says it has nowhere to go.
"""

from __future__ import annotations

from typing import Any, Optional


def issue_deep_link(
    *,
    team_id: Any,
    issue_key: Any,
    step: Optional[int] = None,
    turn: Optional[int] = None,
) -> Optional[str]:
    """``/team/{team_id}/todolist/{issue_key}`` (+ ``?step=&turn=`` when known).

    ``team_id`` is accepted as ``Any`` on purpose: ``issues.team_id`` comes
    back a native int from the issue repository and a string from the
    deliverables join, and both must format to the same URL.

    ``step`` is compared against ``None``, not truth-tested — steps are
    0-based, so ``if step:`` would silently drop every link to a run's first
    step. ``turn`` is read the same way, for the same reason.

    **``turn`` only rides along with a step.** The trajectory keys its nodes by
    the PAIR ``(turn, step)``, so a run that took three turns draws three
    "step 2" nodes and the step alone lands on whichever the page finds first
    (3a Task 8b). Alone, though, a turn names no position — emitting it would
    leave a query key the page can only ignore, so it is dropped with the step.
    """
    if not team_id or not issue_key:
        return None
    url = f"/team/{team_id}/todolist/{issue_key}"
    if step is not None:
        url += f"?step={step}"
        if turn is not None:
            url += f"&turn={turn}"
    return url


__all__ = ["issue_deep_link"]

File: frontend/utils/issueLinks.ts
/**
 * The one frontend builder of an issue's deep link (B7).
 *
 * Mirror of `backend/app/services/issues/issue_links.py::issue_deep_link`,
 * which 3a Task 3b made the single backend builder. The two print the same
 * string for the same row — the backend stamps `deep_link` onto every lineage
 * version and onto the Generated inbox card, and the frontend assembles the
 * same URL wherever it only has a team and a key. A second spelling would
 * drift by a query string and nothing would fail: the link would simply land
 * at the top of the issue instead of on the step you clicked from.
 *
 * **It refuses more often than it builds**, exactly like the backend one.
 * The route only exists inside a team (`router.tsx` registers
 * `todolist/:identifier` under `/team/:teamId` and NOWHERE else — the
 * team-less `/todolist` is a redirect that takes no identifier), so a link
 * built without a team resolves to nothing. `null` is the honest answer, and
 * the caller renders a chip that does not offer the click.
 */

export interface IssueDeepLinkCoords {
  /** 0-based, so it is compared against `null`/`undefined`, never truth-tested. */
  step?: number | null;
  /** Rides along with a step only: alone it names no position, and the page
   *  can only ignore it (same rule as the backend builder). */
  turn?: number | null;
}

export function issueDeepLink(
  teamId: string | number | null | undefined,
  issueKey: string | null | undefined,
  coords: IssueDeepLinkCoords = {},
): string | null {
  // Falsy, not null-ish: the backend's guard is `if not team_id or not
  // issue_key`, which rejects `0` and `""` as well (L1). A team id of 0 is
  // not a team — spelling the refusal differently on the two sides is the
  // drift this mirror exists to prevent.
  if (!teamId || !issueKey) return null;
  let url = `/team/${teamId}/todolist/${issueKey}`;
  const { step, turn } = coords;
  if (step !== null && step !== undefined) {
    url += `?step=${step}`;
    if (turn !== null && turn !== undefined) url += `&turn=${turn}`;
  }
  return url;
}
